modules/apps/roles/roles-test/src/testIntegration/java/com/liferay/roles/service/test/RoleServiceTest.java

```diff
@@ -84,6 +106,55 @@ public void testSearchCount() throws Exception {
 				new LinkedHashMap<>()));
 	}
 
+	@Test
+	public void testSearchWithClassNameId() throws Exception {
+		UserTestUtil.setUser(TestPropsValues.getUser());
+
+		_className = _classNameLocalService.addClassName(
+			RandomTestUtil.randomString());
+
+		_roles.add(
+			_roleService.addRole(
+				_className.getClassName(), RandomTestUtil.nextLong(),
+				RandomTestUtil.randomString(),
+				RandomTestUtil.randomLocaleStringMap(),
+				RandomTestUtil.randomLocaleStringMap(),
+				RoleConstants.TYPE_PROVIDER, null, null));
+		_roles.add(
+			_roleService.addRole(
+				_className.getClassName(), RandomTestUtil.randomLong(),
+				RandomTestUtil.randomString(),
+				RandomTestUtil.randomLocaleStringMap(),
+				RandomTestUtil.randomLocaleStringMap(),
+				RoleConstants.TYPE_PROVIDER, null, null));
+		_roles.add(
+			_roleService.addRole(
+				_className.getClassName(), RandomTestUtil.randomLong(),
+				RandomTestUtil.randomString(),
+				RandomTestUtil.randomLocaleStringMap(),
+				RandomTestUtil.randomLocaleStringMap(),
+				RoleConstants.TYPE_PROVIDER, null, null));
+
+		LinkedHashMap<String, Object> params = new LinkedHashMap<>();
+
+		params.put("classNameId", _className.getClassNameId());
+
+		Assert.assertEquals(
+			_roles.size(),
+			_roleService.searchCount(
+				TestPropsValues.getCompanyId(), StringPool.BLANK,
+				new Integer[] {RoleConstants.TYPE_PROVIDER}, params));
+
+		List<Role> roles = _roleService.search(
+			group.getCompanyId(), StringPool.BLANK,
+			new Integer[] {RoleConstants.TYPE_PROVIDER}, params,
+			QueryUtil.ALL_POS, QueryUtil.ALL_POS, null);
```

Comment:
We may want to consider adding a role that is a provider type, but does _not_ use the class name generated here. That way we can be sure the query isn't just picking up every provider type regardless of className.